joint_trajectory_controller/test/joint_trajectory_controller_test.cpp

```diff
@@ -1289,13 +1300,23 @@ TEST_F(JointTrajectoryControllerTest, goalToleranceViolation)
   EXPECT_TRUE(waitForState(action_client, SimpleClientGoalState::ABORTED, long_timeout));
   EXPECT_EQ(action_client->getResult()->error_code, control_msgs::FollowJointTrajectoryResult::GOAL_TOLERANCE_VIOLATED);
 
+  // Check that we're not moving after restoring perfect control
+  StateConstPtr state1 = getState();
+
   // Restore perfect control
   {
     std_msgs::Float64 smoothing;
     smoothing.data = 0.0;
     smoothing_pub.publish(smoothing);
     ros::Duration(0.5).sleep();
   }
+
+  StateConstPtr state2 =  getState();
+  for (unsigned int i = 0; i < n_joints; ++i)
+  {
+    EXPECT_NEAR(state1->actual.positions[i],     state2->actual.positions[i],     EPS);
```

Comment:
You check it the same way as in the other test.
To rule out other effect, just use the desired values as well.

Reply:
In this case, taking the desired values is not possible. The trajectory has already finished when the goal tolerance is checked. Therefore, desired values do not change anymore.

Reply:
This test should check the behavior of the controller (=desired value). And according to the test output, this code samples the hold trajectory properly.

Reply:
The desired position should match the actual position at the time of the goal tolerance violation. I could rewrite the test to check that.

Reply:
> The desired position should match the actual position at the time of the goal tolerance violation

The final position is based on the hold trajectory -> will move a little bit to obey the speed limits etc.

Reply:
Ah, yes, I forgot about `stop_trajectory_duration`. Do you have an idea how to formulate the test instead?

Reply:
I think you can check it the [same way](https://github.com/ros-controls/ros_controllers/pull/395#discussion_r251198105) in both tests.
But it might be that the difference for the goal check is too small, let's see.

Reply:
For `stop_trajectory_duration == 0`, that test would always pass because the desired state does not change anymore after the tolerance violation.

Reply:
That's why I proposed:
 > I would make the test check state1 velocity and acceleration to be non-zero

Reply:
Before your fix, the desired velocity and acceleration of state1 would ne non-zero, because the controller is still moving towards the goal and motion does not stop.
Your fix make it stop.

The tests must command a goal that takes much longer to reach than the hold trajectory (to stop the motion).

Reply:
I don't think, we are talking about the same problem here. I try to summarize the scenario to clarify the situation:
The robot is moving towards some goal but is delayed and therefore does not follow the trajectory very closely. Now the desired trajectory is finished (desired velocity/acceleration are zero)  and goal tolerances are checked. The robot is still moving as it is delayed.
Previous to my PR the robot would keep moving and reach the goal at an unknown time. My PR changes the behavior so the robot stops moving towards the original goal (as moving there would be useless/dangerous now) and holds its position.

I think what should be tested here, is that the robot stays close to its current position after the goal tolerance violation occurred. I agree, that the desired trajectory should be checked, but simply checking velocity/acceleration is not an option as the original behavior also commands a zero velocity/acceleration in its final state, as the trajectory has finished.

Reply:
> The tests must command a goal that takes much longer to reach than the hold trajectory (to stop the motion).

I thought the same. This could be done in the path-tolerance test case.

I suggest to synchronize with #457 in order to avoid doubling the effort.

@Martin-Oehler Are you still interested in continuing on this?

Reply:
Unfortunately, I am currently occupied by other projects and won't have time to further look into this. Feel free to take over if you are interested.